File: adapter/src/main/java/com/ashlikun/adapter/recyclerview/section/SectionSingAdapter.java
package com.ashlikun.adapter.recyclerview.section;

import android.content.Context;
import android.view.ViewGroup;

import com.ashlikun.adapter.ViewHolder;
import com.ashlikun.adapter.recyclerview.vlayout.SimpleSingAdapter;

import java.util.List;

/**
 * 作者　　: 李坤
 * 创建时间: 9:57 Administrator
 * 邮箱　　：496546144@qq.com
 * <p>
 * 功能介绍：分组adapter 给Vlayout的
 */

public abstract class SectionSingAdapter<T extends SectionEntity> extends SimpleSingAdapter<T> {
    /**
     * 头部ItemType,这里Vlayout是不允许type为负数的
     */
    public static final int TYPE_SECTION = Math.abs("TYPE_SECTION".hashCode());
    /**
     * 头布局id
     */
    protected int mSectionHeadResId = -2;

    public SectionSingAdapter(Context context, List<T> datas) {
        this(context, -DEFAULT_LAYOUT_ID, -2, datas);
    }

    public SectionSingAdapter(Context context, int mSectionHeadResId, List<T> datas) {
        this(context, DEFAULT_LAYOUT_ID, mSectionHeadResId, datas);
    }

    public SectionSingAdapter(Context context, int layoutId, int mSectionHeadResId, List<T> datas) {
        super(context, layoutId, datas);
        this.mSectionHeadResId = mSectionHeadResId;
    }

    @Override
    public int getItemViewType(int position) {
        return getItemData(position).isHeader() ?
                TYPE_SECTION :
                super.getItemViewType(position);
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == TYPE_SECTION) {
            ViewHolder holder = new ViewHolder(mContext, createLayout(parent, mSectionHeadResId, viewType), this);
            setListener(parent, holder, viewType);
            return holder;
        } else {
            return super.onCreateViewHolder(parent, viewType);
        }
    }


    protected boolean isViewTypeHeader(int viewType) {
        return viewType == TYPE_SECTION;
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        if (holder.getItemViewType() == TYPE_SECTION) {
            setListener(recyclerView, holder, holder.getItemViewType());
            bindHeader(holder, getItemData(position));
            return;
        }
        super.onBindViewHolder(holder, position);
    }

    /**
     * 为header设置数据
     *
     * @param holder
     */
    public abstract void bindHeader(ViewHolder holder, T data);
}

File: adapter/src/main/java/com/ashlikun/adapter/recyclerview/section/SectionSingAdapter.kt
package com.ashlikun.adapter.recyclerview.section

import android.content.Context
import android.view.View
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.viewbinding.ViewBinding
import com.ashlikun.adapter.ViewHolder
import com.ashlikun.adapter.recyclerview.AdapterConvert
import com.ashlikun.adapter.recyclerview.SpanSizeLookupGroup
import com.ashlikun.adapter.recyclerview.vlayout.SingAdapter
import com.ashlikun.adapter.recyclerview.vlayout.mode.AdapterBus
import com.ashlikun.adapter.recyclerview.vlayout.mode.LayoutStyle
import kotlin.math.abs

/**
 * 作者　　: 李坤
 * 创建时间: 9:57 Administrator
 * 邮箱　　：496546144@qq.com
 *
 *
 * 功能介绍：分组adapter 给Vlayout的
 */
open class SectionSingAdapter<T : SectionEntity>(
    context: Context,
    initDatas: List<T>? = null,
    //创建ViewBinding的Class,与layoutId 二选一
    override val binding: Class<out ViewBinding>? = null,
    //头布局 ViewBinding
    open val bndingHead: Class<out ViewBinding>? = null,
    //布局文件
    override val layoutId: Int? = null,
    //事件
    override var bus: AdapterBus? = null,
    //头布局id
    open var headLayoutId: Int = View.NO_ID,
    //布局，优先
    layoutStyle: LayoutStyle? = null,
    //ViewType
    override var viewType: Any? = null,
    //转换头
    open var convertHeader: AdapterConvert<T>? = null,
    //初始化的apply 便于执行其他代码
    apply: (SectionSingAdapter<T>.() -> Unit)? = null,
    //转换
    override var convert: AdapterConvert<T>? = null
) : SingAdapter<T>(
    context = context,
    initDatas = initDatas,
    layoutStyle = layoutStyle
) {

    companion object {
        /**
         * 头部ItemType,这里Vlayout是不允许type为负数的
         */
        val TYPE_SECTION = abs("SectionSingAdapter.TYPE_SECTION".hashCode())
    }

    init {
        apply?.invoke(this)
    }

    override fun getItemViewType(position: Int): Int {
        return if (getItemData(position)?.isHeader == true) TYPE_SECTION else super.getItemViewType(
            position
        )
    }

    override fun getLayoutId(viewType: Int): Int? {
        return if (viewType == TYPE_SECTION) return headLayoutId else super.getLayoutId(
            viewType
        )
    }

    override fun getBindClass(viewType: Int): Class<out ViewBinding>? {
        return if (viewType == TYPE_SECTION) return bndingHead else super.getBindClass(
            viewType
        )
    }


    protected fun isPositionHeader(position: Int): Boolean {
        return getItemViewType(position) == TYPE_SECTION
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
        if (recyclerView.layoutManager is GridLayoutManager) {
            val gridLayoutManager = recyclerView.layoutManager as GridLayoutManager
            gridLayoutManager.spanSizeLookup = SpanSizeLookupGroup(gridLayoutManager) {
                isPositionHeader(it)
            }
        }
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        if (holder.itemViewType == TYPE_SECTION) {
            setListener(holder, holder.itemViewType)
            convertHeader?.invoke(holder, getItemData(position))
            return
        }
        super.onBindViewHolder(holder, position)
    }


}